core/src/main/java/org/infinispan/topology/LocalTopologyManagerImpl.java

```diff
@@ -307,13 +307,21 @@ private boolean doHandleTopologyUpdate(String cacheName, CacheTopology cacheTopo
 
          ConsistentHash unionCH = null;
          if (cacheTopology.getPendingCH() != null) {
-            unionCH = cacheStatus.getJoinInfo().getConsistentHashFactory().union(cacheTopology.getCurrentCH(),
-                  cacheTopology.getPendingCH());
+            if (cacheTopology.getPhase() == CacheTopology.Phase.READ_NEW_WRITE_ALL) {
+               // When removing members from topology, we have to make sure that the unionCH has
+               // owners from pendingCH (which is used as the readCH in this phase) before
+               // owners from currentCH, as primary owners must be always readable.
```

Comment:
Will change the comment, but I won't extend the scope of the PR unless necessary.